### PR TITLE
eager qat: refactor conv-bn to make conv with scaled weights a func

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39828 [wip] qat graph mode: add combine_conv_bn pass
* **#39837 eager qat: refactor conv-bn to make conv with scaled weights a func**
* #39840 qat graph mode: make `_forward` of per channel observers scriptable
* #39796 qat graph mode: initial scaffolding for prepare_qat_script

Summary:

WIP, not ready for review yet.

Factors out the logic we will need to keep observation-free in graph mode into
a separate function. This will make it easier to reason about the
equivalent graph mode pass.

No logic or numerics change.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21998753](https://our.internmc.facebook.com/intern/diff/D21998753)

### CODE REVIEW

File: torch/nn/intrinsic/qat/modules/conv_fused.py [hunk 1]
```diff
@@ -79,15 +79,27 @@ def freeze_bn_stats(self):
...
-        scale_factor = self.bn.weight / running_std
-        scaled_weight = self.weight_fake_quant(self.weight * scale_factor.reshape([-1, 1, 1, 1]))
+    # Runs the conv forward with scaled weights and unscales the weights.
+    #
+    # Note: this is implemented as a standalone function to match the
+    #   rules of graph mode quantization, which will mimic this code.
+    #   In particular, the inputs and outputs of this function should be
+    #   observed, and no intermediate values of this function should be
+    #   observed.
+    def _conv_forward_and_unscale(self, input, scaled_weight, scale_factor):
```
Comment (jerryzh168) [inline torch/nn/intrinsic/qat/modules/conv_fused.py:89]: I actually mean an independent function. this is a method for the class

Comment (jerryzh168) [reply torch/nn/intrinsic/qat/modules/conv_fused.py:89]: rule for graph mode right now is we will go through CallMethod, but we will not go through the CallFunction in the list

Comment (vkuzo) [reply torch/nn/intrinsic/qat/modules/conv_fused.py:89]: yep, we're on the same page.  Not intending to directly reuse this code, more structuring the eager mode code the same way we aligned for the graph mode IR, for easier testing during development.

let me mark this PR as wip, not ready for review yet.

File: aten/src/ATen/native/Convolution.cpp [hunk 2]
```diff
@@ -1139,4 +1139,22 @@ std::tuple<Tensor,Tensor,Tensor> _convolution_double_backward(
...
 
+// Performs convolution with a scaled weight, unscales, and returns the
+// unscaled results. This is done for QAT Conv-BN fusion: it contains
+// the logical part of QAT Conv-BN which is after observation (all inputs
+// are observed), and before BN.
+at::Tensor qat_conv2d_and_unscale(
+    const Tensor& input, const Tensor& scaled_weight,
+    const Tensor& scale_factor, const Tensor& bias, IntArrayRef stride,
+    IntArrayRef padding, IntArrayRef dilation, int64_t groups) {
+  Tensor result = at::convolution(
```
Comment (jerryzh168) [inline aten/src/ATen/native/Convolution.cpp:1150]: why are we not using conv2d?

**Review by jerryzh168 — APPROVED** [review]
LG, had one question

File: aten/src/ATen/native/Convolution.cpp [hunk 3]
```diff
@@ -1139,4 +1139,22 @@ std::tuple<Tensor,Tensor,Tensor> _convolution_double_backward(
...
+      input, scaled_weight, bias, stride, padding, dilation, false, {{0, 0}},
+      groups);
+  result = result / scale_factor.reshape({1, -1, 1, 1});
```
Comment (jerryzh168) [inline aten/src/ATen/native/Convolution.cpp:1153]: also should we make this work for conv3d as well?

Comment (jerryzh168) [reply aten/src/ATen/native/Convolution.cpp:1153]: btw fusion is already extended to work with conv3d: https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/fusion.py#L17

File: aten/src/ATen/native/Convolution.cpp [hunk 4]
```diff
@@ -1139,4 +1139,22 @@ std::tuple<Tensor,Tensor,Tensor> _convolution_double_backward(
...
+// Performs convolution with a scaled weight, unscales, and returns the
+// unscaled results. This is done for QAT Conv-BN fusion: it contains
+// the logical part of QAT Conv-BN which is after observation (all inputs
+// are observed), and before BN.
+at::Tensor qat_conv2d_and_unscale(
+    const Tensor& input, const Tensor& scaled_weight,
+    const Tensor& scale_factor, const Tensor& bias, IntArrayRef stride,
+    IntArrayRef padding, IntArrayRef dilation, int64_t groups) {
+  Tensor result = at::convolution(
+      input, scaled_weight, bias, stride, padding, dilation, false, {{0, 0}},
```
Comment (raghuramank100) [inline aten/src/ATen/native/Convolution.cpp:1151]: Are we adding the bias twice here? at::convolution takes in bias as an argument and later bias is added

File: torch/nn/intrinsic/qat/modules/conv_fused.py [hunk 5]
```diff
@@ -83,11 +83,21 @@ def _forward(self, input):
         running_std = torch.sqrt(self.bn.running_var + self.bn.eps)
         scale_factor = self.bn.weight / running_std
         scaled_weight = self.weight_fake_quant(self.weight * scale_factor.reshape([-1, 1, 1, 1]))
-        # this does not include the conv bias
-        conv = self._conv_forward(input, scaled_weight)
-        conv_orig = conv / scale_factor.reshape([1, -1, 1, 1])
-        if self.bias is not None:
-            conv_orig = conv_orig + self.bias.reshape([1, -1, 1, 1])
+
```
Comment (raghuramank100) [inline torch/nn/intrinsic/qat/modules/conv_fused.py:86]: A comment on the original code: I dont follow the logic here: We seem to be calling conv_forward , which includes the bias, though the comment says that "this does not include the conv bias"

Comment (raghuramank100) [reply torch/nn/intrinsic/qat/modules/conv_fused.py:86]: Our original implementation is complicated and can be simplified, we first add a bias as part of the call to _conv_forward, Then we again add the bias to conv_orig. All of this is removed by the batch norm op as part of subtracting the batch mean.

Comment (vkuzo) [reply torch/nn/intrinsic/qat/modules/conv_fused.py:86]: > # this does not include the conv bias

yeah, this comment was not describing the code well

Comment (raghuramank100) [reply torch/nn/intrinsic/qat/modules/conv_fused.py:86]: A cleaner implementation is to do:
``` python
scaled_weight = self.weight_fake_quant(self.weight * scale_factor.reshape([-1, 1, 1, 1]))/scale_factor.reshape([-1,1,1,1])
F.convolution(input, scaled_weight, self.bias, self.stride, padding, self.dilation, self.groups)
```
```

File: aten/src/ATen/native/Convolution.cpp (hunk 4 above)
Comment (raghuramank100) [reply aten/src/ATen/native/Convolution.cpp:1151]: We do not need this op, we can call at::convolution directly, please see below

File: aten/src/ATen/native/Convolution.cpp (hunk 2 above)
Comment (vkuzo) [reply aten/src/ATen/native/Convolution.cpp:1150]: I'm flexible, they eventually call the same thing.  Happy to change it.

File: torch/nn/intrinsic/qat/modules/conv_fused.py (hunk 5 above)
Comment (vkuzo) [reply torch/nn/intrinsic/qat/modules/conv_fused.py:86]: good point.  The code can potentially be made simpler with that refactor, thanks!

from what we discussed in chat, we'd still need a separate function for the above, because we want to insert the fake_quant on conv.weight directly at the op boundary of the aten op (to fit with how the rest of the code works and avoid custom logic).  So, since fake_quant is divided by scale_factor above, it wouldn't work out of the box.

File: aten/src/ATen/native/Convolution.cpp (hunk 3 above)
Comment (vkuzo) [reply aten/src/ATen/native/Convolution.cpp:1153]: yep, for sure.  Planning to do that in future PR, to keep PRs smaller.  Will add a TODO.
